File: docs/07-Graphs-and-Disjoint-Set/cpp/dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <utility>

const int INF = 1e9;

void dijkstra(const std::vector<std::vector<std::pair<int, int>>>& graph, int start, int numVertices) {
    // Min priority queue: pair of (distance, vertex)
    std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<std::pair<int, int>>> pq;
    std::vector<int> distances(numVertices, INF);

    distances[start] = 0;
    pq.push({0, start});

    while (!pq.empty()) {
        int current_dist = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        if (current_dist > distances[u]) continue;

        for (const auto& neighbor : graph[u]) {
            int v = neighbor.first;
            int weight = neighbor.second;

            if (distances[u] + weight < distances[v]) {
                distances[v] = distances[u] + weight;
                pq.push({distances[v], v});
            }
        }
    }

    std::cout << "Shortest distances from vertex " << start << ":\n";
    for (int i = 0; i < numVertices; ++i) {
        std::cout << "To vertex " << i << ": " << distances[i] << "\n";
    }
}

int main() {
    int numVertices = 4;
    std::vector<std::vector<std::pair<int, int>>> graph(numVertices);
    
    // Adjacency list: {neighbor, weight}
    graph[0].push_back({1, 4});
    graph[0].push_back({2, 1});
    graph[1].push_back({3, 1});
    graph[2].push_back({1, 2});
    graph[2].push_back({3, 5});

    dijkstra(graph, 0, numVertices);

    return 0;
}

File: docs/07-Graphs-and-Disjoint-Set/js/dijkstra.js
class PriorityQueue {
    constructor() {
        this.values = [];
    }
    enqueue(val, priority) {
        this.values.push({ val, priority });
        this.sort();
    }
    dequeue() {
        return this.values.shift();
    }
    isEmpty() {
        return this.values.length === 0;
    }
    sort() {
        this.values.sort((a, b) => a.priority - b.priority);
    }
}

function dijkstra(graph, start, numVertices) {
    const distances = Array(numVertices).fill(Infinity);
    const pq = new PriorityQueue();

    distances[start] = 0;
    pq.enqueue(start, 0);

    while (!pq.isEmpty()) {
        const { val: u, priority: dist } = pq.dequeue();

        if (dist > distances[u]) continue;

        const neighbors = graph[u] || [];
        for (const neighbor of neighbors) {
            const { dest: v, weight } = neighbor;

            if (distances[u] + weight < distances[v]) {
                distances[v] = distances[u] + weight;
                pq.enqueue(v, distances[v]);
            }
        }
    }

    return distances;
}

// Driver testing
const graph = {
    0: [{ dest: 1, weight: 4 }, { dest: 2, weight: 1 }],
    1: [{ dest: 3, weight: 1 }],
    2: [{ dest: 1, weight: 2 }, { dest: 3, weight: 5 }],
    3: []
};

const startNode = 0;
const numVertices = 4;
const shortestPaths = dijkstra(graph, startNode, numVertices);

console.log(`Shortest distances from vertex ${startNode}:`);
shortestPaths.forEach((dist, vertex) => {
    console.log(`To vertex ${vertex}: ${dist}`);
});
